### PR TITLE
Iterator pattern

### PR DESCRIPTION
**Pull request by harveyvanveltom**

### CODE REVIEW

File: src/main/java/alien/Alien.java [hunk 1]
```diff
@@ -198,42 +202,48 @@ public void setMovementSpeed(int x) {
      * @return
      */
     public int ifHit(Vector<Bullet> shipBullets) {
-        for (int i = 0; i < shipBullets.size(); i++) {
-        	if (shipBullets.get(i).getY() > ypos-12 && shipBullets.get(i).getY() < ypos + 16) {
-        		if (shipBullets.get(i).getX() > xpos-6 && shipBullets.get(i).getX() < xpos + 16) {
+        Iterator iteralien = ConcreteAggregate.createIterator(shipBullets);
```
Comment (JerPk) [inline src/main/java/alien/Alien.java]: iteralien -> iterShipBullets (better for readability)

File: src/main/java/alien/BossAlien.java [hunk 2]
```diff
@@ -69,123 +70,130 @@ public void vmovement() {
...
     public int ifHit(Vector<Bullet> shipBullets) {
-        for (int i = 0; i < shipBullets.size(); i++) {
-        	if (shipBullets.get(i).getY() > ypos-12 && shipBullets.get(i).getY() < ypos + 20) {
-        		if (shipBullets.get(i).getX() > xpos-6 && shipBullets.get(i).getX() < xpos + 50) {
-                    Game.logfile.writeHit("Alien", xpos, ypos);
+        Iterator iteralien = ConcreteAggregate.createIterator(shipBullets);
 
-                    return i;
+        while (iteralien.hasNext()) {
+            Bullet bullet = (Bullet) iteralien.next();
```
Comment (JerPk) [inline src/main/java/alien/BossAlien.java]: iteralien -> iterShipBullets

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 3]
```diff
@@ -130,61 +129,67 @@ public void init() {
...
+        moveAliens();
     }
 
     public void moveAliens() {
-        for (int i = 0; i < aliens.size(); i++) {
-            Alien alien_obj = (Alien) aliens.get(i);
-            alien_obj.hmovement();
-        }	
-        // this if statement will only be used if all the aliens need to be updated simultaneously.
+        Iterator iteralien = ConcreteAggregate.createIterator(aliens);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: iteralien -> iterAliens

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 4]
```diff
@@ -130,61 +129,67 @@ public void init() {
...
+
+        while (iteralien.hasNext()) {
+            Alien alien = (Alien) iteralien.next();
+            alien.hmovement();
+        }
+
+        // this if statement will only be used if all the aliens need to be
+        // updated simultaneously.
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: Need to be moved horizontally and vertically

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 5]
```diff
@@ -212,47 +217,56 @@ public void listenForKeys() {
...
-        	end();
+            end();
         }
-        for (int i = 0; i < aliens.size(); i++) {
-            int hit = aliens.get(i).ifHit(shipBullets);
-            if (hit != -1) {               
-                if (aliens.get(i).defeated()) {
-                    score = aliens.get(i).addScore(score);
-                    aliens.removeElementAt(i);
+        Iterator iteralien = ConcreteAggregate.createIterator(aliens);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: iteralien -> iterAliens

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 6]
```diff
@@ -212,47 +217,56 @@ public void listenForKeys() {
...
+                    score = alien.addScore(score);
+                    aliens.removeElementAt(iteralien.position());
                 }
                 shipBullets.removeElementAt(hit);
             }
         }
-        for (int i = 0; i < barriers.size(); i++) {
-        	int hit = barriers.get(i).ifHit(alienBullets);
+
+        Iterator iterbarrier = ConcreteAggregate.createIterator(barriers);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: iterbarrier -> iterBarriers

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 7]
```diff
@@ -268,20 +282,30 @@ public void alienShoot() {
      * The method that removes all the bullets that are offscreen.
      */
     public void removeOffScreenBullets() {
-        for (int i = 0; i < alienBullets.size(); i++) {
-            if (alienBullets.get(i).reachedY(450)) {
-                Game.logfile.writeOffscreen("Alien", alienBullets.get(i).getX());
-                alienBullets.removeElementAt(i);
-                i--;
+        Iterator iteralien = ConcreteAggregate.createIterator(alienBullets);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: iteralien -> iterAlienBullets

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 8]
```diff
@@ -268,20 +282,30 @@ public void alienShoot() {
...
             }
+
         }
-        for (int j = 0; j < shipBullets.size(); j++) {
-            if (shipBullets.get(j).getY() <= 0) {
-                Game.logfile.writeOffscreen("Spaceship", shipBullets.get(j).getX());
-                shipBullets.removeElementAt(j);
-                j--;
+
+        Iterator itership = ConcreteAggregate.createIterator(shipBullets);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: itership -> iterShipBullets

File: src/main/java/spaceinvaders/spaceinvaders_framework/Spaceship.java [hunk 9]
```diff
@@ -72,46 +77,50 @@ public Bullet shoot() {
     }
 
     public int ifHit(Vector<Bullet> alienBullets) {
-    	for (int i = 0; i < alienBullets.size(); i++) {
-    	    Bullet testBullet = alienBullets.get(i);
-            if (testBullet instanceof MegaBullet) {
-                if (ifHitMega(alienBullets,i) == true) {
-                    return i;
+        Iterator iteralien = ConcreteAggregate.createIterator(alienBullets);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Spaceship.java]: iteralien -> iterAlienBullets

File: src/main/java/spaceinvaders/spaceinvaders_framework/Game.java [hunk 10]
```diff
@@ -130,61 +129,67 @@ public void init() {
...
+        }
+
+        // this if statement will only be used if all the aliens need to be
+        // updated simultaneously.
         if (Alien.getupdateLogic()) {
-            for (int i = 0; i < aliens.size(); i++) {
-                Alien alien_obj = (Alien) aliens.get(i);
-                alien_obj.vmovement();
+
+            Iterator iteralien2 = ConcreteAggregate.createIterator(aliens);
```
Comment (JerPk) [inline src/main/java/spaceinvaders/spaceinvaders_framework/Game.java]: iteralien2 -> iterAliens2

**Comment by JerPk** [comment]
If the naming of the iterators is fixed it can be merged.

**Comment by Mari3** [comment]
Ok, can be merged

**Pull request merged**